### PR TITLE
Validate Gateway specs attempting to bind with <1024 port without root

### PR DESCRIPTION
**Pull request by howardjohn**
There are 2 commits, the 2nd is just updating golden files since the tests were using old manifests that were binding to port <1024 which then failed the tests

What it looks like for a user:
```
$ istioctl install -f /tmp/iop.yml
Error: failed to install manifests: generated config failed semantic validation: port http2/80 in gateway cluster-local-gateway invalid: targetPort is set to 0, which requires root. Set targetPort to be greater than 1024 or configure values.gateways.istio-ingressgateway.runAsRoot=true
exit status 1
```

### CODE REVIEW

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 1]
```diff
@@ -135,8 +134,67 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
...
 // validateFeatures check whether the config sematically make sense. For example, feature X and feature Y can't be enabled together.
-func validateFeatures(_ *valuesv1alpha1.Values, _ *v1alpha1.IstioOperatorSpec) util.Errors {
-	return nil
+func validateFeatures(values *valuesv1alpha1.Values, spec *v1alpha1.IstioOperatorSpec) util.Errors {
+	return CheckServicePorts(values, spec)
+}
+
+func CheckServicePorts(values *valuesv1alpha1.Values, spec *v1alpha1.IstioOperatorSpec) util.Errors {
+	var errs util.Errors
+	if !values.GetGateways().GetIstioIngressgateway().GetRunAsRoot().GetValue() {
```
Comment (ostromart) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:143]: this long chain of derefs looks worrisome here and below - are we sure there's no null here? if so a comment would be useful about the assumptions.

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 2]
```diff
@@ -135,8 +134,67 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
 }
 
 // validateFeatures check whether the config sematically make sense. For example, feature X and feature Y can't be enabled together.
-func validateFeatures(_ *valuesv1alpha1.Values, _ *v1alpha1.IstioOperatorSpec) util.Errors {
-	return nil
+func validateFeatures(values *valuesv1alpha1.Values, spec *v1alpha1.IstioOperatorSpec) util.Errors {
+	return CheckServicePorts(values, spec)
+}
+
+func CheckServicePorts(values *valuesv1alpha1.Values, spec *v1alpha1.IstioOperatorSpec) util.Errors {
```
Comment (ostromart) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:141]: nit: comment would be good to summarize what's being checked.

File: operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml [hunk 3]
```diff
@@ -8,6 +8,6 @@ spec:
     telemetry:
       v2:
         metadataExchange:
-          wasmEnabled: false
+          wasmEnabled: true
```
Comment (JimmyCYJ) [inline operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml:11]: Is this change related to the binding issue?

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 4]
```diff
@@ -135,8 +134,67 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
...
+	var errs util.Errors
+	if !values.GetGateways().GetIstioIngressgateway().GetRunAsRoot().GetValue() {
+		errs = util.AppendErrs(errs, validateGateways(spec.GetComponents().GetIngressGateways(), "istio-ingressgateway"))
+	}
```
Comment (ostromart) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:145]: should this validate user gateways with different names too? also, i wonder if the name const could be derived rather than hardcoded here.

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 5]
```diff
@@ -135,8 +134,67 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
...
+	if !values.GetGateways().GetIstioEgressgateway().GetRunAsRoot().GetValue() {
+		errs = util.AppendErrs(errs, validateGateways(spec.GetComponents().GetEgressGateways(), "istio-egressgateway"))
+	}
+	for _, port := range values.GetGateways().GetIstioIngressgateway().GetIngressPorts() {
+		var tp int
+		if port["targetPort"] != nil {
+			t, ok := port["targetPort"].(float64)
+			if !ok {
+				continue
```
Comment (ostromart) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:154]: error?

File: operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml (hunk 3 above)
Comment (howardjohn) [reply operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml:11]: No, its related to how the istioctl tests work. See the file name, `data-snapshot` - to avoid having changes to the testdata every time the charts change, we keep a copy of them so its periodically updated. I had to update this copy when making this change or all the tests fail since the old copy is now invalid

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go [hunk 6]
```diff
@@ -135,8 +134,67 @@ func deprecatedSettingsMessage(iop *v1alpha1.IstioOperatorSpec) string {
...
+		for _, p := range gw.GetK8S().GetService().GetPorts() {
+			tp := 0
+			if p.TargetPort != nil && p.TargetPort.Type == intstr.String {
+				// Do not validate named ports
+				continue
+			}
+			if p.TargetPort != nil && p.TargetPort.Type == intstr.Int {
+				tp = int(p.TargetPort.IntVal)
+			}
+			if tp == 0 && p.Port > 1024 {
```
Comment (ostromart) [inline operator/pkg/apis/istio/v1alpha1/validation/validation.go:188]: from here on this seems. a copy of above code - helper function?

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go (hunk 1 above)
Comment (howardjohn) [reply operator/pkg/apis/istio/v1alpha1/validation/validation.go:143]: Get...() is null safe in proto. The functions all look like
```
GetDestination() *Destination {
	if m != nil {
		return m.Destination
	}
	return nil
}
```

File: operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml (hunk 3 above)
Comment (JimmyCYJ) [reply operator/cmd/mesh/testdata/manifest-generate/data-snapshot/profiles/preview.yaml:11]: Thanks!

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go (hunk 4 above)
Comment (howardjohn) [reply operator/pkg/apis/istio/v1alpha1/validation/validation.go:145]: It does, `spec.GetComponents().GetIngressGateways()` is a list of gateways. The name const is referring to the `values` name - NOT the gateway name. runAsRoot is part of values

File: operator/pkg/apis/istio/v1alpha1/validation/validation.go (hunk 6 above)
Comment (howardjohn) [reply operator/pkg/apis/istio/v1alpha1/validation/validation.go:188]: They are different enough that I don't feel its worth it, as both the error message and types are different. It would be more code to make it generic than to copy.

**Review by ostromart — APPROVED** [review]

**Comment by linsun** [comment]
@howardjohn Thanks!  What about upgrade?  Can you do a quick testing with users using istioctl upgrade cmd?

I assume helm users won't get the alert, is this correct?

**Comment by sdake** [comment]
nj @howardjohn!
